torch/csrc/distributed/c10d/ProcessGroupXCCL.cpp

```diff
@@ -0,0 +1,310 @@
+  }
+
+  it = devXCCLCommMap_.find(deviceKey);
+  TORCH_INTERNAL_ASSERT(
+      it != devXCCLCommMap_.end(), "Communicators not populated in cache!");
+
+  return it->second;
+}
+
+template <typename Fn, typename PreProcess, typename PostProcess>
+c10::intrusive_ptr<Work> ProcessGroupXCCL::collective(
+    at::Tensor& input,
+    at::Tensor& output,
+    Fn fn,
+    PreProcess pre,
+    PostProcess post,
+    OpType opType) {
+  using traits = function_traits<Fn>;
+  using attr_t = typename traits::template arg<2>::type;
+  attr_t attr = ccl::create_operation_attr<attr_t>();
+
+  auto device = input.device();
+  const auto key = std::to_string(device.index());
+  auto comm = getXCCLComm(key, device);
+
+  auto stream = xcclStreams_.at(key);
+  std::vector<at::Tensor> inputs{input};
+  std::vector<at::Tensor> outputs{output};
+
+  c10::intrusive_ptr<ProcessGroupXCCL::WorkXCCL> work;
+
+  work = initWork(device, rank_, opType);
+
+  work->outputs_ =
+      std::make_shared<std::vector<at::Tensor>>(std::move(outputs));
+  c10::xpu::XPUCachingAllocator::recordStream(
+      input.storage().data_ptr(), stream);
+  
+  auto ccl_stream = ccl::create_stream(stream.queue());
+  // auto ccl_stream = ccl::create_stream();
+
+  fn(input, output, attr, *comm, ccl_stream);
+
+  work->xcclEndEvent_->record(stream);
+
+  std::vector<c10::Stream> streams = {stream.unwrap()};
+  c10::MultiStreamGuard streamGuard(streams);
+  std::vector<at::Device> devices{device};
+  work->future_ = c10::make_intrusive<at::ivalue::Future>(
+      c10::ListType::create(c10::TensorType::get()), devices);
+  work->future_->markCompleted(at::IValue(*work->outputs_));
+
+  return work;
+}
+
+template <typename Fn>
+c10::intrusive_ptr<Work> ProcessGroupXCCL::collective(
+    at::Tensor& input,
+    at::Tensor& output,
+    Fn fn,
+    OpType opType) {
+  return collective<Fn>(
+      input,
+      output,
+      fn,
+      [](at::xpu::XPUStream&,
+         c10::intrusive_ptr<ProcessGroupXCCL::WorkXCCL>& work) {},
+      [](at::xpu::XPUStream&,
+         c10::intrusive_ptr<ProcessGroupXCCL::WorkXCCL>& work) {},
+      opType);
+}
+
+c10::intrusive_ptr<Work> ProcessGroupXCCL::allreduce(
+    std::vector<at::Tensor>& tensors,
+    const AllreduceOptions& opts) {
+  TORCH_CHECK(
+      tensors.size() == 1, "Expecting one tensor only but got multiple");
+  auto tensor = tensors.back();
+  check_xpu_single_tensor(tensor);
+  if (opts.reduceOp == ReduceOp::AVG) {
+    TORCH_CHECK(false, "Cannot use ReduceOp AVG with XPU")
+  }
+  return collective(
+      tensor,
+      tensor,
+      [&](at::Tensor& input,
+          at::Tensor& output,
+          ccl::allreduce_attr attr,
+          xcclComm_t& comm,
+          ccl::stream& stream) {
+        ccl::event ret_evt;
+        ret_evt = ccl::allreduce(
+            input.data_ptr(),
+            output.data_ptr(),
+            (size_t)input.numel(),
+            getXcclDataType(input.scalar_type()),
+            xcclOps.at(opts.reduceOp),
```

Comment:
There are many reduction types ((beyond scope in torch-ccl) in stock pytorch. So please define a function like `getXcclReduceOp(...)` to achieve reduction type to avoid failures from directly call.